tutorial/sample4/webgl-demo.js

```diff
@@ -1,77 +1,81 @@
+  const gl = canvas.getContext("webgl");
 
-  initWebGL(canvas);      // Initialize the GL context
+  // If we don't have a GL context, give up now
 
-  // Only continue if WebGL is available and working
+  if (!gl) {
+    alert("Unable to initialize WebGL. Your browser or machine may not support it.");
+    return;
+  }
 
-  if (gl) {
-    gl.clearColor(0.0, 0.0, 0.0, 1.0);  // Clear to black, fully opaque
-    gl.clearDepth(1.0);                 // Clear everything
-    gl.enable(gl.DEPTH_TEST);           // Enable depth testing
-    gl.depthFunc(gl.LEQUAL);            // Near things obscure far things
+  // Vertex shader program
 
-    // Initialize the shaders; this is where all the lighting for the
-    // vertices and so forth is established.
+  const vsSource = `
+    attribute vec4 aVertexPosition;
+    attribute vec4 aVertexColor;
 
-    initShaders();
+    uniform mat4 uModelViewMatrix;
+    uniform mat4 uProjectionMatrix;
 
-    // Here's where we call the routine that builds all the objects
-    // we'll be drawing.
+    varying lowp vec4 vColor;
 
-    initBuffers();
+    void main(void) {
+      gl_Position = uProjectionMatrix * uModelViewMatrix * aVertexPosition;
+      vColor = aVertexColor;
+    }
+  `;
 
-    // Set up to draw the scene periodically.
+  // Fragment shader program
 
-    setInterval(drawScene, 15);
-  }
-}
+  const fsSource = `
+    varying lowp vec4 vColor;
 
-//
-// initWebGL
-//
-// Initialize WebGL, returning the GL context or null if
-// WebGL isn't available or could not be initialized.
-//
-function initWebGL() {
-  gl = null;
+    void main(void) {
+      gl_FragColor = vColor;
+    }
+  `;
 
-  try {
-    gl = canvas.getContext("experimental-webgl");
-  }
-  catch(e) {
-  }
+  // Initialize the shaders; this is where all the lighting for the
+  // vertices and so forth is established.
+  const shaderProgram = initShaderProgram(gl, vsSource, fsSource);
 
-  // If we don't have a GL context, give up now
+  // Look up which attribute our shader is using for aVertexPosition
+  const attribLocations = {
+    vertexPosition: gl.getAttribLocation(shaderProgram, "aVertexPosition"),
+    vertexColor: gl.getAttribLocation(shaderProgram, "aVertexColor"),
+  };
 
-  if (!gl) {
-    alert("Unable to initialize WebGL. Your browser may not support it.");
+  // Look up uniform locations
+  const uniformLocations = {
+    projectionMatrix: gl.getUniformLocation(shaderProgram, "uProjectionMatrix"),
+    modelViewMatrix: gl.getUniformLocation(shaderProgram, "uModelViewMatrix"),
+  };
+
+  // Here's where we call the routine that builds all the objects
+  // we'll be drawing.
+  const buffers = initBuffers(gl);
+
+  var then = 0;
+
+  // Draw the scene repeatedly
+  function render(now) {
+    now *= 0.001;  // convert to seconds
+    const deltaTime = now - then;
+    then = now;
+
+    drawScene(gl, shaderProgram, attribLocations, uniformLocations, buffers, deltaTime);
```

Comment:
I like the consistency of collecting the attributes and uniforms on a single object.